### PR TITLE
[WONT MERGE] Fix memory leak in subclass and add more error checking

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64551
* #64746
* #64550

Note that this is not a solution for #64548 which should be addressed as well.

This PR adds two main checks:
-  When creating Tensor pyobjects, we ensure that the type is a subtype of THPVariable (otherwise our pointer casts do very weird stuff).
-  Ensures that we only set the Tensor's pyobj if it is not already set (otherwise we leak the THPVariable). Fix call sites where the user explicitly asked for the pyobject to be changed to properly clear it before setting it again.

### CODE REVIEW

File: test/test_python_dispatch.py [hunk 1]
```diff
@@ -335,6 +336,34 @@ def backward(ctx, grad_output):
...
 $6 = torch._ops.aten.add_($1, $5)''')
 
+    def test_subclass_creation(self):
+        # Make sure these statements runs without error
+        # In particular checking that when internal detach returns
+        # subclasses, these are cleanly overwritten.
+        class Foo(torch.Tensor):
+            pass
+        a = torch.Tensor._make_subclass(Foo, LoggingTensor(torch.rand(2)))
+        b = LoggingTensor(torch.rand(2)).as_subclass(Foo)
```
Comment (ezyang) [inline test/test_python_dispatch.py:346]: Thinking about the preferred behavior here, I kind of want `as_subclass` to hit up `__torch_function__` and the default implementation to raise an error if Foo isn't a subclass of LoggingTensor. Because you have no reason to believe that slicing off all of the LoggingTensor state and turning it into a Foo is a valid operation.

File: test/test_python_dispatch.py [hunk 2]
```diff
@@ -335,6 +336,34 @@ def backward(ctx, grad_output):
...
+
+        # And in case where we don't know if the user wants this subclass
+        # overwritten, raise a nice error.
+        # The standard LoggingTensor will fail because it is not on the right device
+        with self.assertRaisesRegex(TypeError, "expected.*device=cpu.*device=meta"):
```
Comment (ezyang) [inline test/test_python_dispatch.py:351]: TBH, the device error is not very good, we'd prefer the other error

File: torch/csrc/autograd/python_variable.cpp [hunk 3]
```diff
@@ -303,9 +346,12 @@ static PyObject* THPVariable_as_subclass(PyObject* _self, PyObject* args, PyObje
   if (!PyType_Check(cls)) {
     throw torch::TypeError("cls must be a type (got %s)", Py_TYPE(cls)->tp_name);
   }
+  auto data = self.alias();
+  unsafe_clear_pyobj(data);
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:350]: Yeah. So instead of clearing the PyObject, I'd rather just error here (preferably because we attempted to do a `__torch_function__` dispatch and no one said they knew how to do this subclass conversion

File: torch/csrc/autograd/python_variable.cpp [hunk 4]
```diff
@@ -105,6 +105,44 @@ class PyInterpreterHolder {
...
+// For classic binding, this would be as simple as clearing the field and
+// decrefing the pyobj but the special logic that keep Tensor's pyobj alive
+// would prevent this from working as expected.
+// Here we first properly break the link Tensor -> pyobj. Then we break the
+// link pyobj -> Tensor. And finally decref the pyobj.
+//
+// This function assumes that it is used just before calling into
+// init_pyobj() (via THPVariable_NewWithVar) in particular for the interpreter
+// state.
+static void unsafe_clear_pyobj(const Variable& tensor) {
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:119]: I object to this function existing. It breaks the invariant that once a C++ tensor is associated with a PyObject, it is always associated with that PyObject

**Review by ezyang — CHANGES_REQUESTED** [review]
not very hot on the proposed semantics here

File: test/test_python_dispatch.py (hunk 2 above)
Comment (albanD) [reply test/test_python_dispatch.py:351]: The problem is that this error is raised by the tensor options check in the legacy constructor before we get to the actual subclass wrapping.
This is true in general for the legacy Tensor constructor:
```python
import torch

a = torch.rand(2, device="gpu")
b = torch.Tensor(a)
```
This will raise this same device error.

File: torch/csrc/autograd/python_variable.cpp (hunk 3 above)
Comment (albanD) [reply torch/csrc/autograd/python_variable.cpp:350]: But what about subclasses like `nn.Parameter` for which `__torch_function__` is disabled?

File: torch/csrc/autograd/python_variable.cpp (hunk 4 above)
Comment (albanD) [reply torch/csrc/autograd/python_variable.cpp:119]: I agree with that.
I wanted to have a full fix here to be able to drive the discussion.

File: test/test_python_dispatch.py (hunk 1 above)
Comment (albanD) [reply test/test_python_dispatch.py:346]: If we go to `__torch_function__`, how is this default implementation supposed to work? Because even if it is a proper subclass, we will still have the problem that the `.detach()` will generate a Tensor with a pyobj already (of the parent class).
And so if we refuse to "replace" these pyobject, we can't really make such a subclass?

Also, we have some subclass (like nn.Parameter) that disable `__torch_function__` so that might be problematic.

**Comment by ezyang** [comment]
@albanD and I discussed this in workchat and Alban is going to prepare a more comprehensive proposal for how to overhaul these semantics (or at least say why the situation is complicated)
